From: Thomas Renninger <trenn@suse.de>
Subject: x6: Fix reserve_initrd so that acpi_initrd_override is reached
References: bnc#831836
Patch-Mainline: never, code changed and things should be fixed differently

also if initrd is not relocated.

Signed-off-by: Thomas Renninger <trenn@suse.de>

Index: linux-3.7-openSUSE-12.3/arch/x86/kernel/setup.c
===================================================================
--- linux-3.7-openSUSE-12.3.orig/arch/x86/kernel/setup.c
+++ linux-3.7-openSUSE-12.3/arch/x86/kernel/setup.c
@@ -405,13 +405,11 @@ static void __init reserve_initrd(void)
 		 */
 		initrd_start = ramdisk_image + PAGE_OFFSET;
 		initrd_end = initrd_start + ramdisk_size;
-		return;
+	} else {
+		relocate_initrd();
+		memblock_free(ramdisk_image, ramdisk_end - ramdisk_image);
 	}
 
-	relocate_initrd();
-
-	memblock_free(ramdisk_image, ramdisk_end - ramdisk_image);
-
 	acpi_initrd_override((void *)initrd_start, initrd_end - initrd_start);
 }
 #else
